### ISSUE TITLE
Silent data replication when using shard constraints

### ISSUE THREAD

**Issue by vvvm23** [jax-ml/jax#26946]
### Description

When using sharding constraints on intermediate arrays in a `jax.jit` function, the arrays can be silently replicated along that axis, rather than sharded, if the size of the axis is not divisible by the number of devices on the axis.

Reproducing script, taking the example of embedding weights with vocab size `vocab_size`, on 8xGPUs:
```python
import jax
import numpy as np
import jax.numpy as jnp

from jax.sharding import Mesh, PartitionSpec as P, NamedSharding
from jax.lax import with_sharding_constraint as shard

def main():
    vocab_size = 17 # change to 16 for correct sharding
    mesh = Mesh(np.reshape(jax.devices("gpu"), (8,)), ('data',))

    @jax.jit
    def fn():
        embedding = jnp.arange(vocab_size*16).reshape(vocab_size, 16)
        embedding = shard(embedding, P('data', None))
        jax.debug.visualize_array_sharding(embedding)
        return embedding
    
    with mesh:
        embedding = fn().block_until_ready()
    jax.debug.visualize_array_sharding(embedding)

    # uncomment to hard error when not-divisible (expected)
    # embedding = jax.device_put(embedding, NamedSharding(mesh, P('data', None))) 
    # jax.debug.visualize_array_sharding(embedding)

if __name__ == "__main__":
    main()
```

When `vocab_size` is not divisible by 8, the array sharding visualisation will show `embedding` actually gets fully replicated with no error or warning. This is unexpected as the `device_put` function will hard error when attempting to do a similar thing.

When it is divisible by 8, it will correctly replicate the array.

The expected behaviour is to throw an error when attempting to do this type of sharding. Otherwise you could get a silent increase in the memory requirements of jit functions (eg. imagine a vocab size of >1M elements)

### System info (python version, jaxlib version, accelerator, etc.)

```
jax:    0.5.0
jaxlib: 0.5.0
numpy:  2.2.3
python: 3.10.12 (main, Feb  4 2025, 14:57:36) [GCC 11.4.0]
device info: NVIDIA <redacted>, 8 local devices"
process_count: 1
platform: <redacted>

$ nvidia-smi
<redacted>
```

**Comment by yashk2810** [comment]
That's not silent, it's working as intended.

Uneven sharding works for intermediates but if you return that intermediate with uneven sharding, we'll fully replicate it. I admit this is not written down anywhere.

Once JAX supports uneven sharding at the top level, this will work as you expect.

**Comment by vvvm23** [comment]
I understand JAX/XLA cannot support uneven (ragged?) shardings. It makes sense in the SPMD paradigm.

However, to fallback silently to replication here, I think is unexpected behaviour? It should throw a hard error like we see for `jax.device_put`.

**Comment by vvvm23** [comment]
TIL about the intermediate case! 

But even so, this is a common way to generate sharded data, but it fails silently. Wouldn't it be reasonable to detect and catch this kind of case?

**Comment by yashk2810** [comment]
No, these are the semantics of with_sharding_constraint.

You can use jit(f, out_shardings=...) to get the error

**Comment by vvvm23** [comment]
Can you point me to where these semantics are defined? 

This seems to contradict the docs for the op that state "This is a strict constraint for the GSPMD partitioner and not a hint". Does "strict" meaning something different here?

**Comment by yashk2810** [comment]
> Can you point me to where these semantics are defined?

It's not officially written down but I can send a change to add it to the docs.

**Comment by vvvm23** [comment]
I guess another doc change could be to `jax.jit`:

```
out_shardings (Any) – optional, a Sharding or pytree with Sharding leaves and structure that is a tree prefix of the output of fun. If provided, it has the same effect as applying corresponding jax.lax.with_sharding_constraint`s to the output of ``fun`().
```

As you said, using `jax.jit(..., out_shardings)` instead should induce the error (I have not tested) but the doc seems to imply they should have identical effects.
